(CLOUD-160) Add subnet support to instance

garethr opened the pull request:
Cleaned up commits in https://github.com/puppetlabs/puppetlabs-aws/pull/48 and opened this as an alternative.

Iristyle commented:
Much easier to understand 0f5cab8 now - thanks.  Will review later today.

Iristyle commented:
Supercedes #48,  #34 and #23 :icecream:

daveseff commented:
Any idea when this PR will get merged in?

cmurphy commented on lib/puppet/provider/ec2_instance/v2.rb:
```diff
@@ -105,13 +130,57 @@ def create
+
+        if subnet.nil?
+          raise Puppet::Error,
+            "Invalid security groups '#{groups.join(', ')}' not found in VPCs '#{vpc_groups.keys.join(', ')}'"
```
This would be more friendly if it raised an error earlier than this if vpc_groups is empty. Currently, it claims the group was "`not found in VPCs ''`" which is hard to read.

cmurphy replied:
This also seems to be a misleading message, as sometimes the error occurs if the subnet is invalid but the security group is valid.

garethr replied:
I've added the additional check and altered the mentioned error message to be more factual.

garethr commented:
@Iristyle @cmurphy I've just rebased this against master and made the error handling improvement suggested.

garethr commented:
@daveseff sorry for the delay. Lots going on and a bunch of travelling last week has meant a bit of a backlog. If you get chance to try the patch out do feed back, we'll try and get it into master soon.

Iristyle commented:
Hey @garethr looks like we need another rebase against master... but might want to wait until we get #63 and #46 merged first.

cmurphy commented:
The acceptance tests fail for me when VPC is optional (i.e. when using the standard puppetlabs account):

```
Failures:

  1) ec2_loadbalancer should create a new load balancer with the specified name
     Failure/Error: loadbalancers = @aws.get_loadbalancers(name)
     Aws::ElasticLoadBalancing::Errors::LoadBalancerNotFound:
       Cannot find Load Balancer colleentakd2278737c2994faeb1e34
     # ./spec/spec_helper_acceptance.rb:107:in `get_loadbalancers'
     # ./spec/acceptance/loadbalancer_spec.rb:15:in `get_loadbalancer'
     # ./spec/acceptance/loadbalancer_spec.rb:55:in `block (3 levels) in <top (required)>'

  2) ec2_loadbalancer should create a new load balancer with the specified availability zone
     Failure/Error: loadbalancers = @aws.get_loadbalancers(name)
     Aws::ElasticLoadBalancing::Errors::LoadBalancerNotFound:
       Cannot find Load Balancer colleentakd2278737c2994faeb1e34
     # ./spec/spec_helper_acceptance.rb:107:in `get_loadbalancers'
     # ./spec/acceptance/loadbalancer_spec.rb:15:in `get_loadbalancer'
     # ./spec/acceptance/loadbalancer_spec.rb:55:in `block (3 levels) in <top (required)>'

  3) ec2_loadbalancer should create a new load balancer not part of a VPC
     Failure/Error: loadbalancers = @aws.get_loadbalancers(name)
     Aws::ElasticLoadBalancing::Errors::LoadBalancerNotFound:
       Cannot find Load Balancer colleentakd2278737c2994faeb1e34
     # ./spec/spec_helper_acceptance.rb:107:in `get_loadbalancers'
     # ./spec/acceptance/loadbalancer_spec.rb:15:in `get_loadbalancer'
     # ./spec/acceptance/loadbalancer_spec.rb:55:in `block (3 levels) in <top (required)>'

  4) ec2_loadbalancer should create a new load balancer with one associated instance
     Failure/Error: loadbalancers = @aws.get_loadbalancers(name)
     Aws::ElasticLoadBalancing::Errors::LoadBalancerNotFound:
       Cannot find Load Balancer colleentakd2278737c2994faeb1e34
     # ./spec/spec_helper_acceptance.rb:107:in `get_loadbalancers'
     # ./spec/acceptance/loadbalancer_spec.rb:15:in `get_loadbalancer'
     # ./spec/acceptance/loadbalancer_spec.rb:55:in `block (3 levels) in <top (required)>'

  5) ec2_loadbalancer should create a new load balancer with no associated security groups
     Failure/Error: loadbalancers = @aws.get_loadbalancers(name)
     Aws::ElasticLoadBalancing::Errors::LoadBalancerNotFound:
       Cannot find Load Balancer colleentakd2278737c2994faeb1e34
     # ./spec/spec_helper_acceptance.rb:107:in `get_loadbalancers'
     # ./spec/acceptance/loadbalancer_spec.rb:15:in `get_loadbalancer'
     # ./spec/acceptance/loadbalancer_spec.rb:55:in `block (3 levels) in <top (required)>'
```

and 21 failures when using the VPC-only account, most of which look like 

```
Failures:

  1) ec2_instance should create a new instance with the specified name
     Failure/Error: expect(instances.count).to eq(1)

       expected: 1
            got: 0

       (compared using ==)
     # ./spec/acceptance/instance_spec.rb:15:in `get_instance'
     # ./spec/acceptance/instance_spec.rb:36:in `block (3 levels) in <top (required)>'
```

Does this depend on other acceptance test changes in other PRs? Or the VPC PR (#43)?

garethr commented:
@cmurphy are the load balancer test failures repeatable? ie. if you run multiple times do you always see failures? What about if you run that spec file on it's own? I've just run them a few times and been unable to recreate. Thanks.

cmurphy commented:
@garethr ah, you're right that it seems to be a transient error not related to this change.

garethr commented:
@cmurphy I've rebased this against master, and fixed the directly related (ie. instance subnet) issues on the VPC only account.

More generally I'm looking at describe_account_attributes and using the information from that to split the logic up more clearly. When I do that I'll pull the implementation apart here a bit and take things out of the now too large create method.
